nni/compression/experiment/config/utils.py

```diff
@@ -0,0 +1,185 @@
+            assert isinstance(module_type, str)
+            _module_types.add(module_type)
+
+    # unfold module types as module names, add them to summary
+    for module_name, module in model.named_modules():
+        module_type = type(module).__name__
+        if module_type in _module_types:
+            module_names_summary.add(module_name)
+        _all_module_names.add(module_name)
+
+    # add module names to summary
+    if module_names:
+        for module_name in module_names:
+            if module_name not in _all_module_names:
+                # need warning, module_name not exist
+                continue
+            else:
+                module_names_summary.add(module_name)
+
+    # remove module names in exclude_module_names from module_names_summary
+    if exclude_module_names:
+        for module_name in exclude_module_names:
+            if module_name not in _all_module_names:
+                # need warning, module_name not exist
+                continue
+            if module_name in module_names_summary:
+                module_names_summary.remove(module_name)
+
+    return list(module_names_summary)
+
+
+def generate_compression_search_space(config: CompressionConfig, vessel: CompressionVessel) -> Dict[str, Dict]:
+    """
+    Using config (constraints & priori) and vessel (model-related) to generate the hpo search space.
+    """
+
+    search_space = {}
+    model, _, evaluator, dummy_input, _, _, _, _ = vessel.export()
+    flops, params, results = count_flops_params(model, dummy_input, verbose=False, mode='full')
+    metric = evaluator(model)
+
+    module_names_summary = _summary_module_names(model, config.module_types, config.module_names, config.exclude_module_names)
+    for module_name in module_names_summary:
+        search_space['{}{}'.format(KEY_MODULE_NAME, module_name)] = {'_type': 'uniform', '_value': [0, 1]}
+
+    assert not config.pruners or not config.quantizers
+
+    # TODO: hard code for step 1, need refactor
+    search_space[KEY_PRUNERS] = {'_type': 'choice', '_value': [pruner_config.json() for pruner_config in config.pruners]}
+
+    original_target = {'flops': flops, 'params': params, 'metric': metric, 'results': results}
+
+    # TODO: following fucntion need improvement
+    flops_theta = _flops_theta_helper(config.flops, flops)
+    params_theta = _flops_theta_helper(config.params, params)
+    metric_theta = _metric_theta_helper(config.metric, metric)
+    thetas = {'flops': flops_theta, 'params': params_theta, 'metric': metric_theta}
+
+    search_space[KEY_VESSEL] = {'_type': 'choice', '_value': [vessel.json()]}
+    search_space[KEY_ORIGINAL_TARGET] = {'_type': 'choice', '_value': [original_target]}
+    search_space[KEY_THETAS] = {'_type': 'choice', '_value': [thetas]}
+    return search_space
+
+
+def parse_params(kwargs: Dict[str, Any]) -> Tuple[Dict[str, str], List[Dict[str, Any]], CompressionVessel, Dict[str, Any], Dict[str, Any]]:
```

Comment:
maybe we can define classes for the returned objects, which will make typing more clear

Reply:
good suggestion, will follow up